File: Android applications/p0461expandablelistevents/java/com/korobeynikov/p0461expandablelistevents/AdapterHelper.java
package com.korobeynikov.p0461expandablelistevents;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

import android.content.Context;
import android.widget.ArrayAdapter;
import android.widget.SimpleExpandableListAdapter;

public class AdapterHelper {

    final String ATTR_GROUP_NAME="groupName";
    final String ATTR_PHONE_NAME="phoneName";
    String[] groups={"HTC","Samsung","LG"};
    String[] phonesHTC={"Sensation","Desire","Wildfire","Hero"};
    String[] phonesSams={"Galaxy S II","Galaxy Nexus","Wave"};
    String[] phonesLG={"Optimus","Optimus Link","Optimus Black","Optimus One"};
    ArrayList<Map<String,String>> groupData;
    ArrayList<Map<String,String>> childDataItem;
    ArrayList<ArrayList<Map<String,String>>> childData;
    Map<String,String> m;
    Context ctx;

    AdapterHelper(Context _ctx){
        ctx=_ctx;
    }

    SimpleExpandableListAdapter adapter;

    SimpleExpandableListAdapter getAdapter(){
        groupData=new ArrayList<Map<String,String>>();
        for(String group:groups){
            m=new HashMap<String,String>();
            m.put(ATTR_GROUP_NAME,group);
            groupData.add(m);
        }
        String[] groupFrom={ATTR_GROUP_NAME};
        int[] groupTo={android.R.id.text1};
        childData=new ArrayList<ArrayList<Map<String,String>>>();

        childDataItem=new ArrayList<Map<String,String>>();
        for(String phone:phonesHTC){
            m=new HashMap<String,String>();
            m.put(ATTR_PHONE_NAME,phone);
            childDataItem.add(m);
        }
        childData.add(childDataItem);

        childDataItem=new ArrayList<Map<String,String>>();
        for(String phone:phonesSams){
            m=new HashMap<String,String>();
            m.put(ATTR_PHONE_NAME,phone);
            childDataItem.add(m);
        }
        childData.add(childDataItem);

        childDataItem=new ArrayList<Map<String,String>>();
        for(String phone:phonesLG){
            m=new HashMap<String,String>();
            m.put(ATTR_PHONE_NAME,phone);
            childDataItem.add(m);
        }
        childData.add(childDataItem);

        String[] childFrom={ATTR_PHONE_NAME};
        int[] childTo={android.R.id.text1};
        adapter=new SimpleExpandableListAdapter(ctx, groupData, android.R.layout.simple_expandable_list_item_1,
                                                groupFrom, groupTo, childData, android.R.layout.simple_list_item_1,
                                                childFrom,childTo);
        return adapter;
    }

    String getGroupText(int groupPos){
        return ((Map<String,String>)(adapter.getGroup(groupPos))).get(ATTR_GROUP_NAME);
    }

    String getChildText(int groupPos,int childPos){
        return ((Map<String,String>)(adapter.getChild(groupPos,childPos))).get(ATTR_PHONE_NAME);
    }

    String getGroupChildText(int groupPos,int childPos){
        return getGroupText(groupPos)+" "+getChildText(groupPos,childPos);
    }
}

File: KotlinLessons/AndroidLessons/p0461expandablelistevents/java/ru/korobeynikov/p0461expandablelistevents/AdapterHelper.kt
package ru.korobeynikov.p0461expandablelistevents

import android.content.Context
import android.widget.SimpleExpandableListAdapter

class AdapterHelper {

    val ATTR_GROUP_NAME="groupName"
    val ATTR_PHONE_NAME="phoneName"
    val groups=arrayOf("HTC","Samsung","LG")
    val phonesHTC=arrayOf("Sensation","Desire","Wildfire","Hero")
    val phonesSams=arrayOf("Galaxy S II","Galaxy Nexus","Wave")
    val phonesLG=arrayOf("Optimus","Optimus Link","Optimus Black","Optimus One")
    lateinit var groupData: ArrayList<Map<String,String>>
    lateinit var childDataItem: ArrayList<Map<String,String>>
    lateinit var childData: ArrayList<ArrayList<Map<String,String>>>
    lateinit var m: Map<String,String>
    lateinit var ctx: Context

    constructor(_ctx:Context){
        ctx=_ctx
    }

    lateinit var adapter: SimpleExpandableListAdapter

    @JvmName("getAdapter1")
    fun getAdapter():SimpleExpandableListAdapter{
        groupData=ArrayList<Map<String,String>>()
        for(group in groups){
            m=HashMap<String,String>()
            (m as HashMap<String,String>).put(ATTR_GROUP_NAME,group)
            groupData.add(m)
        }
        val groupFrom=arrayOf(ATTR_GROUP_NAME)
        val groupTo=arrayOf(android.R.id.text1)
        childData=ArrayList<ArrayList<Map<String,String>>>()
        childDataItem=ArrayList<Map<String,String>>()
        for(phone in phonesHTC){
            m=HashMap<String,String>()
            (m as HashMap<String,String>).put(ATTR_PHONE_NAME,phone)
            childDataItem.add(m)
        }
        childData.add(childDataItem)
        childDataItem=ArrayList<Map<String,String>>()
        for(phone in phonesSams){
            m=HashMap<String,String>()
            (m as HashMap<String,String>).put(ATTR_PHONE_NAME,phone)
            childDataItem.add(m)
        }
        childData.add(childDataItem)
        childDataItem=ArrayList<Map<String,String>>()
        for(phone in phonesLG){
            m=HashMap<String,String>()
            (m as HashMap<String,String>).put(ATTR_PHONE_NAME,phone)
            childDataItem.add(m)
        }
        childData.add(childDataItem)
        val childFrom=arrayOf(ATTR_PHONE_NAME)
        val childTo=arrayOf(android.R.id.text1)
        adapter= SimpleExpandableListAdapter(ctx,groupData,android.R.layout.simple_expandable_list_item_1,
            groupFrom,groupTo.toIntArray(),childData,android.R.layout.simple_list_item_1,childFrom,childTo.toIntArray())
        return adapter
    }

    fun getGroupText(groupPos:Int): String? {
        return (adapter.getGroup(groupPos) as Map<String,String>).get(ATTR_GROUP_NAME)
    }

    fun getChildText(groupPos:Int,childPos:Int): String? {
        return (adapter.getChild(groupPos,childPos) as Map<String,String>).get(ATTR_PHONE_NAME)
    }

    fun getGroupChildText(groupPos:Int,childPos:Int):String{
        return getGroupText(groupPos)+" "+getChildText(groupPos,childPos)
    }
}